pallets/xcm-core-buyer/src/lib.rs

```diff
@@ -0,0 +1,494 @@
+        /// Orchestartor chain `ParaId`. Used in `absolute_multilocation` to convert the
+        /// `interior_multilocation` into what the relay chain needs to allow to `DepositAsset`.
+        type SelfParaId: Get<ParaId>;
+        type RelayChain: Default
+            + Encode
+            + Decode
+            + TypeInfo
+            + EncodeLike
+            + Clone
+            + PartialEq
+            + sp_std::fmt::Debug;
+        /// Limit how many in-flight XCM requests can be sent to the relay chain in one block.
+        type MaxParathreads: Get<u32>;
+        /// Get the parathread params. Used to verify that the para id is a parathread.
+        // TODO: and in the future to restrict the ability to buy a core depending on slot frequency
+        type GetParathreadParams: GetParathreadParams;
+        /// Get a list of collators assigned to this parathread. Used to verify the collator proof.
+        type GetAssignedCollators: GetParathreadCollators<Self::AccountId>;
+        /// A configuration for base priority of unsigned transactions.
+        ///
+        /// This is exposed so that it can be tuned for particular runtime, when
+        /// multiple pallets send unsigned transactions.
+        #[pallet::constant]
+        type UnsignedPriority: Get<TransactionPriority>;
+
+        type WeightInfo: WeightInfo;
+    }
+
+    #[pallet::event]
+    #[pallet::generate_deposit(pub(super) fn deposit_event)]
+    pub enum Event<T: Config> {
+        /// An XCM message to buy a core for this parathread has been sent to the relay chain.
+        BuyCoreXcmSent { para_id: ParaId },
+    }
+
+    #[pallet::error]
+    pub enum Error<T> {
+        InvalidProof,
+        ErrorValidatingXCM,
+        ErrorDeliveringXCM,
+        /// An order for this para id already exists
+        OrderAlreadyExists,
+        /// The para id is not a parathread
+        NotAParathread,
+        /// There are too many in-flight orders, buying cores will not work until some of those
+        /// orders finish.
+        InFlightLimitReached,
+        /// There are no collators assigned to this parathread, so no point in buying a core
+        NoAssignedCollators,
+        /// This collator is not assigned to this parathread
+        CollatorNotAssigned,
+        /// The `XcmWeights` storage has not been set. This must have been set by root with the
+        /// value of the relay chain xcm call weight and extrinsic weight
+        XcmWeightStorageNotSet,
+        /// Converting a multilocation into a relay relative multilocation failed
+        ReanchorFailed,
+    }
+
+    /// Proof that I am a collator, assigned to a para_id, and I can buy a core for that para_id
+    #[derive(Encode, Decode, CloneNoBound, PartialEq, Eq, DebugNoBound, TypeInfo)]
+    #[scale_info(skip_type_params(T))]
+    pub struct BuyCoreCollatorProof<T: Config> {
+        account: T::AccountId,
+        // TODO
+        _signature: (),
```

Comment:
This signature is required, since we are spending tokens from the parathread tank account right?

Reply:
I think we decided not to make this PR bigger with the validation of the signature but @tmpolaczyk correct me if I am wrong. Also in that case let's update the TODO

Reply:
Yes, the validation logic will be a separate PR so that it can be reviewed better